src/gadgets/curves/bn256/ec_pairing.rs

```diff
@@ -0,0 +1,556 @@
+use std::sync::Arc;
+
+use pairing::bn256::{Fq2, FROBENIUS_COEFF_FQ6_C1, XI_TO_Q_MINUS_1_OVER_2};
+
+use crate::{
+    cs::traits::cs::ConstraintSystem,
+    gadgets::{boolean::Boolean, curves::SmallField, non_native_field::traits::NonNativeField},
+};
+
+use super::*;
+
+// Curve parameter for the BN256 curve
+const CURVE_U_PARAMETER: u64 = 4965661367192848881;
+const SIX_U_PLUS_TWO_WNAF: [i8; 65] = [
+    0, 0, 0, 1, 0, 1, 0, -1, 0, 0, 1, -1, 0, 0, 1, 0, 0, 1, 1, 0, -1, 0, 0, 1, 0, -1, 0, 0, 0, 0,
+    1, 1, 1, 0, 0, -1, 0, 0, 1, 0, 0, 0, 0, 0, -1, 0, 0, 1, 1, 0, 0, -1, 0, 0, 0, 1, 1, 0, -1, 0,
+    0, 1, 0, 1, 1,
+];
+
+/// Struct for the line function evaluation for the BN256 curve (addition and doubling).
+/// The line function is used in the Miller loop of the pairing function.
+pub struct LineFunctionEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    c0: BN256Fq2NNField<F>,
+    c3: BN256Fq2NNField<F>,
+    c4: BN256Fq2NNField<F>,
+    point: BN256SWProjectivePointTwisted<F>,
+    _marker: std::marker::PhantomData<CS>,
+}
+
+impl<F, CS> LineFunctionEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    /// Creates a zero instance of the line function evaluation for the BN256 curve.
+    pub fn zero(cs: &mut CS, params: &Arc<BN256BaseNNFieldParams>) -> Self {
+        Self {
+            c0: BN256Fq2NNField::zero(cs, params),
+            c3: BN256Fq2NNField::zero(cs, params),
+            c4: BN256Fq2NNField::zero(cs, params),
+            point: BN256SWProjectivePointTwisted::zero(cs, params),
+            _marker: std::marker::PhantomData::<CS>,
+        }
+    }
+
+    /// Returns the point of the line function evaluation.
+    pub fn point(&self) -> BN256SWProjectivePointTwisted<F> {
+        self.point.clone()
+    }
+
+    /// Returns the coefficients of the line function evaluation.
+    pub fn c0c3c4(&self) -> (BN256Fq2NNField<F>, BN256Fq2NNField<F>, BN256Fq2NNField<F>) {
+        (self.c0.clone(), self.c3.clone(), self.c4.clone())
+    }
+
+    /// This function conducts the doubling step in the Miller loop for the BN256 curve.
+    /// Namely, given `Q` in `E'(Fp2)` and `P` in `E(Fp)`, it computes the line function
+    /// together with the resultant point `T=2*Q`. The implementation is based
+    /// on the _Algorithm 26_ from https://eprint.iacr.org/2010/354.pdf.
+    pub fn doubling_step(
+        cs: &mut CS,
+        q: &mut BN256SWProjectivePointTwisted<F>,
+        p: &mut BN256SWProjectivePoint<F>,
+    ) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        // 1. tmp0 <- X_Q^2; 2. tmp1 <- Y_Q^2; 3. tmp2 <- tmp1^2;
+        let mut tmp0 = q.x.square(cs);
+        tmp0.normalize(cs);
```

Comment:
do we need this many calls to normalize? i think the nonnative field gadget can take a few overflows before needing to reduce again right?

Reply:
Hey, @jules, the excessive amount of normalization calls is reasoned by the desire to conduct extra check that pairing has no problems. Currently, I am trying to reduce this amount, meanwhile guaranteeing that implementation is still 100% correct. For example, I removed all normalization calls from the doubling step, and all the tests still pass. The issue which I am worried about is that there won't be such an input, which would require normalization, which is not needed for another input. Basically, static checks that the implementations is legit, agnostic of the input. So, in a nutshell, the question is how to understand where to put normalization calls - is it achievable by experimenting in comp-time?

Reply:
hey @NikitaMasych ! you can ensure that the input passed to the function is reduced beforehand by calling `enforce_reduced`. this should guarantee that your input can always take the exact same amount of overflows before giving an error. hope that helps!